nemo/utils/debug_hook.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import torch
+
+
+def get_forward_hook(name, trainer, rank, logger, dump_to_file=False):
+    """
+    A forward hook to dump all of the module input and output norms. It is called at every time after forward() has computed an output.
+    Only float type input/output tensor norms are computed.
+    For more details about the forward hook, check https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html
+
+    Args:
+        name: tensor name
+        trainer: PTL trainer
+        rank: worker rank
+        logger: PTL log function
+        dump_to_file:  wether dump the csv file to the disk
+    """
+    if dump_to_file:
+        os.makedirs('debug_info', exist_ok=True)
+        fp = open(f'debug_info/forward_{name}_rank{rank}.txt', 'w')
+        header = False
+
+    def forward_hook(module, inputs, outputs):
+        nonlocal header
```

Comment:
!?? Why is such hack needed ?

Reply:
hook function doesn't have states. The only state is coming from the enclousre. I want to make sure the csv file has the header which is generated for this first time invoke the hook function.   Do you have a better solution for this?

Reply:
I don't have a better solution other than using a higher order function to proper enclose this nonlocal variable. But I dunno how that interacts with such hooks. I just consider this a dangerous pattern